crates/types/src/testing/querier.rs

```diff
@@ -0,0 +1,256 @@
+}
+
+impl MockQuerier {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn with_info(mut self, info: InfoResponse) -> Self {
+        self.info = Some(info);
+        self
+    }
+
+    pub fn with_app_config<K, V>(mut self, key: K, value: V) -> StdResult<Self>
+    where
+        K: Into<String>,
+        V: Serialize,
+    {
+        let key = key.into();
+        let value = value.to_json_value()?;
+
+        self.app_configs.insert(key, value);
+        Ok(self)
+    }
+
+    pub fn with_balance<T>(mut self, address: Addr, denom: &str, amount: T) -> Self
+    where
+        T: Into<Uint256>,
+    {
+        self.balances
+            .entry(address)
+            .or_default()
+            .insert(denom.to_string(), amount.into());
+        self
+    }
+
+    pub fn with_supplies<T>(mut self, denom: &str, amount: T) -> Self
+    where
+        T: Into<Uint256>,
+    {
+        self.supplies.insert(denom.to_string(), amount.into());
+        self
+    }
+
+    pub fn with_code<T>(mut self, code: T) -> Self
+    where
+        T: Into<Binary>,
+    {
+        let code = code.into();
+        let code_hash = code.hash256();
+
+        self.codes.insert(code_hash, code);
+        self
+    }
+
+    pub fn with_contract(mut self, address: Addr, contract: ContractInfo) -> Self {
+        self.contracts.insert(address, contract);
+        self
+    }
+
+    pub fn with_raw_query_handler(mut self, handler: RawQueryHandler) -> Self {
+        self.raw_query_handler = Some(handler);
+        self
+    }
+
+    pub fn with_smart_query_handler(mut self, handler: SmartQueryHandler) -> Self {
+        self.smart_query_handler = Some(handler);
+        self
+    }
+}
+
+impl Querier for MockQuerier {
+    fn query_chain(&self, req: Query) -> StdResult<QueryResponse> {
+        match req {
+            Query::Info {} => {
+                let info = self.info.clone().expect("[MockQuerier]: info is not set");
+                Ok(QueryResponse::Info(info))
+            },
+            Query::AppConfig { key } => {
+                let value = self
+                    .app_configs
+                    .get(&key)
+                    .cloned()
+                    .ok_or_else(|| StdError::data_not_found::<Json>(key.as_bytes()))?;
+                Ok(QueryResponse::AppConfig(value))
+            },
+            Query::AppConfigs { start_after, limit } => {
+                // Using the `BTreeMap::range` method is more efficient, but for
+                // testing purpose this is good enough.
+                let entries = self
+                    .app_configs
+                    .iter()
+                    .filter(|(k, _)| {
+                        if let Some(lower_bound) = &start_after {
+                            *k > lower_bound
+                        } else {
+                            true
+                        }
+                    })
+                    .take(limit.unwrap_or(u32::MAX) as usize)
```

Comment:
I don't like this... Better if we introduce a trait extension that adds a `may_take` method that takes a `Option<T>` where `T: Into<usize>`